File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack=new Stack<>();

        char[] arr=s.toCharArray();

        for(char ch:arr){
            if(ch=='{' || ch=='(' || ch=='['){
                stack.push(ch);
            }else{
                if(stack.size()==0) return false;
                char opp=stack.peek();
                if(ch=='}' && opp=='{'){
                    stack.pop();
                }else if(opp=='(' && ch==')'){
                    stack.pop();
                }else if(opp=='[' && ch==']'){
                    stack.pop();
                }else{
                    return false;
                }
            }
        } 
        if(stack.size()!=0) return false;
        return true;
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        openbrac=set(['(','{','['])
        closebrac=set([')','}',']'])

        map={
            '{':'}',
            '(':')',
            '[':']'
        }

        stack=[]

        for x in s:
            if x in openbrac:
                stack.append(x)
            else:
                if not stack:
                    return False
                if map[stack[-1]]==x:
                    stack.pop()
                else:
                    return False
        if stack:
            return False
        return True
                